### Môn Lập trình cho khoa học dữ liệu
#### Giảng viên: Trần Trung Kiên


<h1><center>Đồ án cuối kỳ</center></h1>

MSSV |	Họ và tên | Tên tài khoản github
--- | --- | ---
19120186 |	Đỗ Lê Khánh Đăng | dolekhanhdang
19120412 |	Nguyễn Minh Tú   |  tuak123ax

### Table of Contents
* [1. Về dữ liệu](#1)
* [2. Khám phá dữ liệu](#2)
    * [2.1. Các thư viện cần thiết](#2.1)
    * [2.2. Đọc dữ liệu](#2.2)
    * [2.3. Dữ liệu có bao nhiêu dòng và bao nhiêu cột?](#2.3)
    * [2.4. Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?](#2.4)
    * [2.5. Dữ liệu có các dòng bị lặp không?](#2.5)
    * [2.6. Mỗi cột có ý nghĩa gì?](#2.6)
    * [2.7. Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?](#2.7)
    * [2.8. Cột có dtype là object nghĩa là sao?](#2.8)
    * [2.9. Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?](#2.9)
    * [2.10. Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?](#2.10)
* [3. Các câu hỏi ý nghĩa cần trả lời](#3)
    * [Câu hỏi 1: Những công việc nào có mức lương trung bình cao nhất?](#3.1)
    * [Câu hỏi 2: Trong tất cả các kỹ năng(Python,spark,aws,excel...), những kỹ năng nào là quan trọng nhất?](#3.2)
    * [Câu hỏi 3: Mức lương của công ty có phụ thuộc vào size của công ty hay không, liệu có những khác nhau về kỹ năng đối với từng loại công việc giữa các size công ty không? ](#3.3)
* [4. Nhìn lại quá trình làm đồ án](#4)
* [5. Tài liệu tham kháo:](#5)

### 1. Về dữ liệu <a class="anchor" id="1"></a>
Dữ liệu này về lương của những công việc liên quan đến data: data scientist, data engineer, .... <br />
Nguồn: https://www.kaggle.com/nikhilbhathi/data-scientist-salary-us-glassdoor. <br />
Trong mục mô tả của dữ liệu:<br />
- The dataset is scraped from the Glassdoor website using Selenium scrapper. After scrapping, the raw dataset was cleaned and made usable for performing data analysis and modelling. The dataset contains information about the minimum salary, maximum salary, average salary, job description, age of the company in years, etc.
- Dữ liệu được lấy bằng cách scraped từ website của Glassdoor với Selenium scrapper. Sau khi lấy, dữ liệu đã được xử lý phù hợp cho phân tích dữ liệu và mô hình hóa. Tập dữ liệu chứa thông tin về mức lương tối thiểu, tối đa, trung bình, mô tả công việc, tuổi đời công ty,..
- License: CC0 Public Domain

### 2. Khám phá dữ liệu <a class="anchor" id="2"></a>

#### 2.1. Các thư viện cần thiết <a class="anchor" id="2.1"></a>

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

#### 2.2. Đọc dữ liệu <a class="anchor" id="2.2"></a>

In [2]:
data = pd.read_csv('data_cleaned_2021.csv',index_col='index')
data.head()

Job Title              Salary Estimate  \
index                                                           
0                 Data Scientist   $53K-$91K (Glassdoor est.)   
1      Healthcare Data Scientist  $63K-$112K (Glassdoor est.)   
2                 Data Scientist   $80K-$90K (Glassdoor est.)   
3                 Data Scientist   $56K-$97K (Glassdoor est.)   
4                 Data Scientist  $86K-$143K (Glassdoor est.)   

                                         Job Description  Rating  \
index                                                              
0      Data Scientist\r\nLocation: Albuquerque, NM\r\...     3.8   
1      What You Will Do:\r\n\r\nI. General Summary\r\...     3.4   
2      KnowBe4, Inc. is a high growth information sec...     4.8   
3      *Organization and Job ID**\r\nJob ID: 310709\r...     3.8   
4      Data Scientist\r\nAffinity Solutions / Marketi...     2.9   

                                       Company Name         Location  \
index                                                                  
0                          Tecolote Research\r\n3.8  Albuquerque, NM   
1      University of Maryland Medical System\r\n3.4    Linthicum, MD   
2                                    KnowBe4\r\n4.8   Clearwater, FL   
3                                       PNNL\r\n3.8     Richland, WA   
4                         Affinity Solutions\r\n2.9     New York, NY   

         Headquarters          Size  Founded   Type of ownership  ... tensor  \
index                                                             ...          
0          Goleta, CA   501 - 1000      1973   Company - Private  ...      0   
1       Baltimore, MD       10000+      1984  Other Organization  ...      0   
2      Clearwater, FL   501 - 1000      2010   Company - Private  ...      0   
3        Richland, WA  1001 - 5000      1965          Government  ...      0   
4        New York, NY     51 - 200      1998   Company - Private  ...      0   

      hadoop tableau bi  flink  mongo  google_an   job_title_sim  \
index                                                              
0          0       1  1      0      0          0  data scientist   
1          0       0  0      0      0          0  data scientist   
2          0       0  0      0      0          0  data scientist   
3          0       0  0      0      0          0  data scientist   
4          0       0  0      0      0          0  data scientist   

       seniority_by_title Degree  
index                             
0                      na      M  
1                      na      M  
2                      na      M  
3                      na     na  
4                      na     na  

[5 rows x 41 columns]

#### 2.3. Dữ liệu có bao nhiêu dòng và bao nhiêu cột? <a class="anchor" id="2.3"></a>

In [3]:
num_rows = data.shape[0]
num_cols = data.shape[1]
print('Num rows =',num_rows)
print('Num cols =',num_cols)

Num rows = 742
Num cols = 41


Dữ liệu có 742 dòng và 42 cột

#### 2.4. Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không? <a class="anchor" id="2.4"></a>

Như trong mô tả, mỗi dòng là một công việc liên quan đến data và các thông tin liên quan đến công việc đó.

#### 2.5. Dữ liệu có các dòng bị lặp không? <a class="anchor" id="2.5"></a>

In [4]:
data.duplicated().sum()

275

Hmm, vậy là có 275 dòng bị lặp. Việc dể lại cũng không mang lợi ích gì nên bây giờ nên loại chúng luôn.

In [5]:
data = data.drop_duplicates()
data.duplicated().sum()

0

#### 2.6. Mỗi cột có ý nghĩa gì? <a class="anchor" id="2.6"></a>

In [6]:
data.columns

Index(['Job Title', 'Salary Estimate', 'Job Description', 'Rating',
       'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors',
       'Hourly', 'Employer provided', 'Lower Salary', 'Upper Salary',
       'Avg Salary(K)', 'company_txt', 'Job Location', 'Age', 'Python',
       'spark', 'aws', 'excel', 'sql', 'sas', 'keras', 'pytorch', 'scikit',
       'tensor', 'hadoop', 'tableau', 'bi', 'flink', 'mongo', 'google_an',
       'job_title_sim', 'seniority_by_title', 'Degree'],
      dtype='object')

- Job Title : Tên công việc, vd như data scientist, junior data scientist, senior data scientist, etc ... <br />
- Salary Estimate: Khoảng lương và nguồn lấy được.<br />
- Job Description: Mô tả công việc, công ty mong muốn gì từ vị trí này.<br />
- Rating: Rating của công ty.<br />
- Company Name: Tên công ty.<br />
- Location: Địa điểm tuyển dụng.<br />
- Headquarters: Địa chỉ trụ sở công ty.<br />
- Size: Ước tính số lượng nhân viên của công ty.<br />
- Founded: Năm công ty được thành lập.<br />
- Type of ownership: Công ty thuộc chính phủ, tư nhân hay các loại khác.<br />
- Industry: Công ty hoạt động chủ yếu trong ngành công nghiệp nào, IT industry, Pharmaceutical industry , etc...<br />
- Sector: Lĩnh vực mà công ty hoạt động.<br />
- Revenue: Tổng lợi nhuận của công ty mỗi năm. <br />
- Competitors:  Các đối thủ cạnh tranh hiện tại của công ty.<br />
- Hourly: 1 nếu lương này là theo giờ, 0 là các loại khác. <br />
- Employer provided: Mức lương được cung cấp bởi công ty hay không? <br />
- Lower Salary: Mức lương tối thiểu được báo cáo . <br />
- Upper Salary: Mức lương tối đa được báo cáo cho .<br />
- Avg Salary(K): Mức lương trung bình .<br />
- company_txt: chứa tên công ty. <br />
- Job Location: Bang của nơi tuyển dụng.<br />
- Age: Tuổi của công ty, tính bằng năm. <br />
- Python:  Nếu kỹ năng Python là bắt buộc thì là 1. Còn lại là 0.<br />
- spark: Nếu kỹ năng Spark là bắt buộc thì là 1. Còn lại là 0.<br />
- aws: Nếu kỹ năng aws là bắt buộc thì là 1. Còn lại là 0.<br />
- excel: Nếu kỹ năng excel là bắt buộc thì là 1. Còn lại là 0.<br />
- sql: Nếu kỹ năng SQL là bắt buộc thì là 1. Còn lại là 0.<br />
- sas: Nếu kỹ năng SAS là bắt buộc thì là 1. Còn lại là 0.<br />
- keras: Nếu kỹ năng Keras là bắt buộc thì là 1. Còn lại là 0.<br />
- pytorch: Nếu kỹ năng Pytorch là bắt buộc thì là 1. Còn lại là 0.<br />
- scikit: Nếu kỹ năng Scikit là bắt buộc thì là 1. Còn lại là 0.<br />
- tensor: Nếu kỹ năng Tensor là bắt buộc thì là 1. Còn lại là 0.<br />
- hadoop: Nếu kỹ năng Hadoop là bắt buộc thì là 1. Còn lại là 0.<br />
- tableau: Nếu kỹ năng tableau là bắt buộc thì là 1. Còn lại là 0.<br />
- bi: Nếu kỹ năng PowerBi là bắt buộc thì là 1. Còn lại là 0.<br />
- flink: Nếu kỹ năng Flink là bắt buộc thì là 1. Còn lại là 0.<br />
- mongo: Nếu kỹ năng MongoDB là bắt buộc thì là 1. Còn lại là 0.<br />
- google_an: Nếu chứng chỉ Google analytics là bắt buộc thì là 1. Còn lại là 0.<br />
- job_title_sim: Đơn giản hóa tên công việc. <br />
- seniority_by_title: Thâm niên . <br />
- Degree: 
M: Nếu chức danh công việc yêu cầu bằng cấp hoặc cung cấp số năm kinh nghiệm để có được nó. P: Nếu chức danh công việc yêu cầu bằng cấp hoặc cung cấp số năm kinh nghiệm. <br />

#### 2.7. Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không? <a class="anchor" id="2.7"></a>

In [7]:
data.dtypes

Job Title              object
Salary Estimate        object
Job Description        object
Rating                float64
Company Name           object
Location               object
Headquarters           object
Size                   object
Founded                 int64
Type of ownership      object
Industry               object
Sector                 object
Revenue                object
Competitors            object
Hourly                  int64
Employer provided       int64
Lower Salary            int64
Upper Salary            int64
Avg Salary(K)         float64
company_txt            object
Job Location           object
Age                     int64
Python                  int64
spark                   int64
aws                     int64
excel                   int64
sql                     int64
sas                     int64
keras                   int64
pytorch                 int64
scikit                  int64
tensor                  int64
hadoop                  int64
tableau   

Có vẻ không có vấn đề gì với kiểu dữ liệu của các cột.

#### 2.8. Cột có dtype là object nghĩa là sao? <a class="anchor" id="2.8"></a>

Trong Pandas, kiểu dữ liệu object thường ám chỉ chuỗi, nhưng thật ra kiểu dữ liệu object có thể chứa một đối tượng bất kỳ trong Python (vì thật ra ở bên dưới kiểu dữ liệu object chứa địa chỉ). Nếu một cột trong dataframe có dtype là object thì có thể các phần tử trong cột này sẽ có kiểu dữ liệu khác nhau; để biết được kiểu dữ liệu thật sự của các phần tử trong cột này thì ta phải truy xuất vào từng phần tử và dùng câu lệnh type để xem kiểu dữ liệu.

In [8]:
def open_object_dtype(s):
    dtypes = set()

    s=s.apply(type)
    dtypes.update(s.unique().tolist())
    
    return dtypes

In [9]:
list_numerics = data.describe().columns.tolist()
list_cate = [ index for index in data.columns.tolist() if index not in list_numerics]
for index in list_cate:
    print(open_object_dtype(data[index]))

{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}
{<class 'str'>}


Vậy là các cột đều có kiểu dữ liệu là chuỗi, chúng ta không cần xử lý gì thêm.

#### 2.9. Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào? <a class="anchor" id="2.9"></a>

In [10]:
data.describe()

Rating      Founded      Hourly  Employer provided  Lower Salary  \
count  467.000000   467.000000  467.000000         467.000000    467.000000   
mean     3.651820  1832.441113    0.021413           0.021413     74.783726   
std      0.805355   508.661732    0.144913           0.144913     29.920107   
min     -1.000000    -1.000000    0.000000           0.000000     15.000000   
25%      3.400000  1939.000000    0.000000           0.000000     53.500000   
50%      3.700000  1991.000000    0.000000           0.000000     71.000000   
75%      4.100000  2007.000000    0.000000           0.000000     90.500000   
max      5.000000  2019.000000    1.000000           1.000000    202.000000   

       Upper Salary  Avg Salary(K)         Age      Python       spark  ...  \
count    467.000000     467.000000  467.000000  467.000000  467.000000  ...   
mean     128.312634     101.548180   45.605996    0.556745    0.237687  ...   
std       43.941982      36.381891   52.491641    0.497302    0.426123  ...   
min       16.000000      15.500000   -1.000000    0.000000    0.000000  ...   
25%       98.000000      76.250000   11.000000    0.000000    0.000000  ...   
50%      124.000000      97.500000   25.000000    1.000000    0.000000  ...   
75%      150.000000     121.000000   56.000000    1.000000    0.000000  ...   
max      306.000000     254.000000  277.000000    1.000000    1.000000  ...   

            keras     pytorch      scikit      tensor      hadoop     tableau  \
count  467.000000  467.000000  467.000000  467.000000  467.000000  467.000000   
mean     0.040685    0.051392    0.079229    0.102784    0.175589    0.211991   
std      0.197772    0.221032    0.270386    0.304002    0.380878    0.409157   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

               bi       flink       mongo   google_an  
count  467.000000  467.000000  467.000000  467.000000  
mean     0.083512    0.017131    0.051392    0.017131  
std      0.276951    0.129897    0.221032    0.129897  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000    0.000000  
75%      0.000000    0.000000    0.000000    0.000000  
max      1.000000    1.000000    1.000000    1.000000  

[8 rows x 24 columns]

Giá trị max của các cột thì không có gì bất thường, tuy nhiên các giá trị min có vài chỗ là -1 như Rating,Age,... Lý do bởi vì trong lúc thu thập dữ liệu, những chỗ nào không có hay không thu nhập được thì tác giả sẽ để là -1. Như vậy để lúc sau tính tỷ lệ giá trị thiếu, ta sẽ thay những giá trị = -1 bằng np.nan

In [11]:
data = data.replace(-1,np.nan)
data = data.replace('-1',np.nan)
data.describe()

Rating      Founded      Hourly  Employer provided  Lower Salary  \
count  460.000000   434.000000  467.000000         467.000000    467.000000   
mean     3.722609  1971.850230    0.021413           0.021413     74.783726   
std      0.568729    52.794282    0.144913           0.144913     29.920107   
min      1.900000  1744.000000    0.000000           0.000000     15.000000   
25%      3.400000  1958.750000    0.000000           0.000000     53.500000   
50%      3.750000  1995.000000    0.000000           0.000000     71.000000   
75%      4.100000  2008.000000    0.000000           0.000000     90.500000   
max      5.000000  2019.000000    1.000000           1.000000    202.000000   

       Upper Salary  Avg Salary(K)         Age      Python       spark  ...  \
count    467.000000     467.000000  434.000000  467.000000  467.000000  ...   
mean     128.312634     101.548180   49.149770    0.556745    0.237687  ...   
std       43.941982      36.381891   52.794282    0.497302    0.426123  ...   
min       16.000000      15.500000    2.000000    0.000000    0.000000  ...   
25%       98.000000      76.250000   13.000000    0.000000    0.000000  ...   
50%      124.000000      97.500000   26.000000    1.000000    0.000000  ...   
75%      150.000000     121.000000   62.250000    1.000000    0.000000  ...   
max      306.000000     254.000000  277.000000    1.000000    1.000000  ...   

            keras     pytorch      scikit      tensor      hadoop     tableau  \
count  467.000000  467.000000  467.000000  467.000000  467.000000  467.000000   
mean     0.040685    0.051392    0.079229    0.102784    0.175589    0.211991   
std      0.197772    0.221032    0.270386    0.304002    0.380878    0.409157   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

               bi       flink       mongo   google_an  
count  467.000000  467.000000  467.000000  467.000000  
mean     0.083512    0.017131    0.051392    0.017131  
std      0.276951    0.129897    0.221032    0.129897  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000    0.000000  
75%      0.000000    0.000000    0.000000    0.000000  
max      1.000000    1.000000    1.000000    1.000000  

[8 rows x 24 columns]

Giá trị min đã trở lại bình thường. Giờ ta sẽ xem tỷ lệ thiếu của các cột.

In [12]:
data[list_numerics].isnull().sum()

Rating                7
Founded              33
Hourly                0
Employer provided     0
Lower Salary          0
Upper Salary          0
Avg Salary(K)         0
Age                  33
Python                0
spark                 0
aws                   0
excel                 0
sql                   0
sas                   0
keras                 0
pytorch               0
scikit                0
tensor                0
hadoop                0
tableau               0
bi                    0
flink                 0
mongo                 0
google_an             0
dtype: int64

Chỉ có 3 cột có giá trị thiếu là Rating, Founded, Age

#### 2.10. Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào? <a class="anchor" id="2.10"></a>

Ở các cột categorical thì những giá trị thiếu được biểu diễn bằng chuỗi 'na' (not available). Vậy ra sẽ thay các chuỗi đó bằng np.nan trước khi tính các giá trị thiếu.

In [13]:
data = data = data.replace('na',np.nan)

In [14]:
data[list_cate].isnull().sum()

Job Title               0
Salary Estimate         0
Job Description         0
Company Name            0
Location                0
Headquarters            1
Size                    0
Type of ownership       0
Industry                7
Sector                  7
Revenue                 0
Competitors           285
company_txt             0
Job Location            0
job_title_sim           6
seniority_by_title    340
Degree                245
dtype: int64

Competitors và Headquarters, Sector, Industry, seniority_by_title, Degree đêu có giá trị thiếu 

Số lượng các giá trị khác nhau

In [15]:
data[list_cate].nunique()

Job Title             264
Salary Estimate       416
Job Description       463
Company Name          343
Location              200
Headquarters          197
Size                    8
Type of ownership       9
Industry               59
Sector                 24
Revenue                13
Competitors           127
company_txt           343
Job Location           37
job_title_sim           9
seniority_by_title      2
Degree                  2
dtype: int64

In [16]:
for col in list_cate:
    print(data[col].unique()[:3])

['Data Scientist' 'Healthcare Data Scientist' 'Research Scientist']
['$53K-$91K (Glassdoor est.)' '$63K-$112K (Glassdoor est.)'
 '$80K-$90K (Glassdoor est.)']
['Data Scientist\r\nLocation: Albuquerque, NM\r\nEducation Required: Bachelor’s degree required, preferably in math, engineering, business, or the sciences.\r\nSkills Required:\r\nBachelor’s Degree in relevant field, e.g., math, data analysis, database, computer science, Artificial Intelligence (AI); three years’ experience credit for Master’s degree; five years’ experience credit for a Ph.D\r\nApplicant should be proficient in the use of Power BI, Tableau, Python, MATLAB, Microsoft Word, PowerPoint, Excel, and working knowledge of MS Access, LMS, SAS, data visualization tools, and have a strong algorithmic aptitude\r\nExcellent verbal and written communication skills, and quantitative analytical skills are required\r\nApplicant must be able to work in a team environment\r\nU.S. citizenship and ability to obtain a DoD Secret Clea

Có vẻ không có vấn đề gì

### 3. Các câu hỏi ý nghĩa cần trả lời <a class="anchor" id="3"></a>

#### Câu hỏi 1: Những công việc nào có mức lương trung bình cao nhất? <a class="anchor" id="3.1"></a>

Ý nghĩa: trả lời được câu hỏi này sẽ giúp ta biết được công việc nào sẽ có mức lương thế nào để sau này có thể chọn cho mình một công việc có mức lương mà mình mong muốn.

#### Khám phá và xử lý dữ liệu để trả lời

Chúng ta sẽ dựa vào dữ liệu ở cột 'Job Title' và 'Avg Salary(K)' để trả lời câu hỏi này.

#### Bước 1:

Dùng hàm groupby để gom nhóm dataframe ban đầu theo tên công việc và dùng hàm mean() để tính mức lương trung bình của công việc đó rồi lưu vào 'avgSalary_df'

In [17]:
avgSalary_df=data.groupby(by='Job Title').mean()

#### Bước 2:

Tạo series 'result_series' trong đó index là tên công việc và value là lương trung bình của công việc đó(sort giá trị theo thứ tự từ lớn đến bé để dễ quan sát) 

In [18]:
result_series=avgSalary_df['Avg Salary(K)'].sort_values(ascending=False)
result_series

Job Title
Director II, Data Science - GRM Actuarial                 254.0
Principal Machine Learning Scientist                      232.5
Principal Data Scientist with over 10 years experience    225.0
Lead Data Engineer                                        205.0
Director II, Data Science - GRS Predictive Analytics      194.5
                                                          ...  
Medical Technologist / Clinical Laboratory Scientist       41.5
Data Analytics Manager                                     40.5
Project Scientist                                          39.5
Associate Data Analyst                                     38.5
Senior Operations Data Analyst, Call Center Operations     27.5
Name: Avg Salary(K), Length: 264, dtype: float64

In [19]:
len(result_series)

264

Như ta thấy thì 'result_series' chứa đến 264 phần tử.

Do đó khi ta vẽ biểu đồ sẽ rất khó quan sát.

Vậy nên chúng ta sẽ bỏ bớt các phần tử có giá trị nhỏ hơn hoặc bằng mean của series.

In [ ]:
result_series=result_series[result_series > result_series.mean()]
len(result_series)

121

Vẫy còn 121 phần tử (khá nhiều).

Chúng ta tiếp tục bỏ bớt cho tới khi nào còn khoảng 30 giá trị thì dừng.

In [ ]:
result_series=result_series[result_series > result_series.mean()]
len(result_series)

In [ ]:
result_series=result_series[result_series > result_series.mean()]
len(result_series)

#### Bước 4:

Từ kết quả ở trên, vẽ barh chart, trong đó trục hoành là mức lương trung bình và trục tung là tên công việc.

Đặt tên trục hoành là 'AVG Salary' và tên trục tung là 'Job Title'

In [ ]:
result_df=result_series.to_frame()
result_df.columns=['AVG Salary']
result_df.plot.barh(ylabel='Job Title',figsize=(20,15))

Từ biểu đồ trên ta có thể trả lời câu hỏi:

-Những công việc có mức lương cao nhất là: Director II, Data Science - GRM Actuarial ; Principal Machine Learning Scientist ; Principal Data Scientist with over 10 years experience...

-Vậy những công việc có mức lương cao nhất là những công việc ở vị trí cao như giám đốc, quản lý... hay những công việc cần nhiều kinh nghiệm(>10 năm). Từ đó sinh viên sẽ có mục tiêu rõ ràng hơn để hướng đến.

#### Câu hỏi 2: Trong tất cả các kỹ năng(Python,spark,aws,excel...), những kỹ năng nào là quan trọng nhất? <a class="anchor" id="3.2"></a>

Ý nghĩa: trả lời được câu hỏi này sẽ giúp ta biết được những kỹ năng cần thiết cho công việc để có thể trau dồi một cách chuẩn xác nhất.

#### Khám phá và xử lý dữ liệu để trả lời

Chúng ta sẽ dựa vào các cột từ vị trí 22 tới vị trí 37 để trả lời câu hỏi này. Vì các cột này là các cột cho ta biết kỹ năng đó có cần thiết cho công việc hay ko.

#### Bước 1:

Lưu danh sách kỹ năng cần thiết vào 'skill_list'.

In [ ]:
skill_list=data.columns[22:38].tolist()
skill_list

#### Bước 2:

Ứng với mỗi kỹ năng, ta sẽ tính số lượng công việc cần tới kỹ năng này và lưu vào series 'result_skill' với index là tên kỹ năng và value là số lượng công việc cần đến kỹ năng này.

In [ ]:
def Calculate_Num_Needed_Skill(skill_list):
    result_sr=pd.Series(index=skill_list,dtype='float64')
    for skill in skill_list:
        num_yes=data[skill].value_counts()[1]
        result_sr[skill]=num_yes
    return result_sr

In [ ]:
result_skill=Calculate_Num_Needed_Skill(skill_list)
result_skill=result_skill.sort_values(ascending=False)
result_skill

#### Bước 3:

Từ kết quả ở trên, vẽ barh chart, trong đó trục hoành là số lượng công việc cần đến kỹ năng này và trục tung là tên kỹ năng.

Đặt tên trục hoành là 'Num_Job_Needed' và tên trục tung là 'Skill'

In [ ]:
result_skill_df=result_skill.to_frame()
result_skill_df.columns=['Num_Job_Needed']
result_skill_df.plot.barh(ylabel='Skill',figsize=(20,15))

Từ biểu đồ trên ta có thể trả lời câu hỏi:

-Những kỹ năng quan trọng nhất là: Python, sql, excel (số lượng công ty cần các kỹ năng này bỏ xa số lượng công ty cần các kỹ năng khác).

-Từ câu hỏi này ta có thể thấy Python là quan trọng nhất đối với một nhà khoa học dữ liệu. Do đó chúng ta nên rèn luyện kỹ năng Python để sau này có thể kiếm được một công việc ổn định, đồng thời rèn luyện thêm kiến thức về sql và cả excel nữa.

#### Câu hỏi 3: Mức lương của công ty có phụ thuộc vào size của công ty hay không, liệu có những khác nhau về kỹ năng đối với từng loại công việc giữa các size công ty không? <a class="anchor" id="3.3"></a>
Trả lời câu hỏi này sẽ cho ta một cái nhìn toàn cảnh về tình hình lương của các công ty lớn nhỏ, sự khác biệt về yêu cầu của các  loại công ty để định hướng cho phù hợp. Nếu bạn muốn làm những công ty lớn, môi trường chuyên nghiệp thì bạn cần chú trọng về gì. Hoặc bạn muốn công ty startup, môi trường năng động thì cần gì.
- Bước 1: Tiền xử lý dữ liệu 
    + Các giá trị đã được tiền xử lý sẵn nên không cần làm gì thêm. 
- Bước 2: Tìm câu trả lời

In [ ]:
temp = data.groupby(['Size','job_title_sim'])['Avg Salary(K)'].mean().reset_index()
temp = temp.pivot(index = 'Size', columns = 'job_title_sim', values = 'Avg Salary(K)')
temp.style.highlight_max(axis=0,color = 'pink')

Như ta thấy, không phải công ty có size càng lớn thì trung bình lương càng cao. <br />
Lương trung bình cao nhất của 3 nghề analyst, data engineer, data modeler đều tập trung ở công ty có size từ 51-200. Lương trung bình cao nhất của 2 nghề data analitics, machine learning engineer thì ở công ty có size 201 - 500. <br />
Tiếp theo ta sẽ xét xem liệu có những khác nhau về kỹ năng đối với từng loại công việc giữa các size công ty không?

In [ ]:
skill_name = ['Python',
       'spark', 'aws', 'excel', 'sql', 'sas', 'keras', 'pytorch', 'scikit',
       'tensor', 'hadoop', 'tableau', 'bi', 'flink', 'mongo', 'google_an' ]
analyst_data = data.groupby(['Size','job_title_sim'])[skill_name].sum().reset_index()
value_size = data.groupby(['Size','job_title_sim']).size().reset_index(name='counts')

In [ ]:
for job_name in analyst_data['job_title_sim'].unique().tolist():  
    visual_1 = analyst_data[analyst_data['job_title_sim'] == job_name][skill_name].values.tolist()
    name_ = analyst_data[analyst_data['job_title_sim'] == job_name]["Size"].values.tolist()
    size_ = value_size[value_size['job_title_sim']==job_name]['counts'].values.tolist()
    fig = go.Figure()
    sum_1 = [0]*len(skill_name)
    sum_2 = [0]*len(skill_name)
    size_1 = 0
    size_2 = 0 
    for index in range(len(visual_1)):
        if name_[index] in ['1 - 50 ', '201 - 500 ', '51 - 200 ','501 - 1000 ']:
            sum_1 = [sum(x) for x in zip(visual_1[index], sum_1)]
            size_1 += size_[index]
        elif name_[index] != 'unknown': 
            sum_2 = [sum(x) for x in zip(visual_1[index], sum_1)]
            size_2 += size_[index]

    fig.add_trace(go.Scatterpolar(
                  r=[(x / size_1)*100 for x in sum_1],
                  theta=skill_name,
                  fill='toself',
                  name= '<=1000'
            ))
    fig.add_trace(go.Scatterpolar(
                  r=[(x / size_2)*100 for x in sum_2],
                  theta=skill_name,
                  fill='toself',
                  name= '>1000'
            ))
    fig.update_layout(
      polar=dict(
        radialaxis=dict(
          visible=True,
          range=[0, 100]
        )),
      showlegend=True,
        title=job_name
    )

    fig.show()



Để dễ so sánh ta sẽ phân size công ty thành 2 loại, ít hơn 1000 nhân viên và nhiều hơn 1000 nhân viên.<br />
Như ta thấy, các công ty >1000 có xu hướng yêu cầu chung về một hoặc nhiều loại kỹ năng gì đó cho công việc. Ví dụ như ở machine learning engineer, data analitics, Data scientist project manager có một nhóm các kỹ năng mà 100% các công ty >1000 đều yêu cầu phải có. Xu hướng này ít hơn ở bên các công ty <1000 nhân viên. 
Một đặc điểm ta nhận ra nữa là cùng một công việc nhưng công ty <1000 nhân viên lại yêu cầu nhiều kỹ năng hơn, như vậy có thể bạn sẽ làm nhiều nhiệm vụ khác nhau, phát triển nhiều thứ. Còn ở công ty lớn thì bạn sẽ tập trung làm những kỹ năng cốt lõi và rèn luyện chúng tốt hơn.

### 4. Nhìn lại quá trình làm đồ án <a class="anchor" id="4"></a>


#### Trong quá trình làm, nhóm đã gặp không ít khó khăn:
- Khánh Đăng:
    + Việc đầu tiên là tìm câu hỏi có ý nghĩa để trả lời và tìm data set để trả lời nó. Đáng lẽ là nên tìm câu hỏi trước -> tìm data nhưng lúc đầu nhóm đã làm ngược lại, nhóm tìm ra data về các lần phóng, về tên lửa của SpaceX nhưng nhóm nhận ra không đủ kiến thức để phân tích dữ liệu thiên về một chuyên ngành riêng biệt vậy nên nhóm đã chuyển chủ đề về lương của các nghề thuộc lĩnh vực data.
    + Khi visualize để tìm câu trả lời cho câu hỏi 3, lúc đầu nhóm visualize tất cả các size công ty chung với nhau rất là hỗn loạn không nhìn ra gì, sau nhóm mới chia ra là công ty <= 1000 và >1000 nhân viên.
<br />
- Minh Tú :
     + Khó khăn trong việc dùng git, đặc biệt phần xử lý conflict lúc merge. 
     + Trong lúc tìm câu hỏi ý nghĩa cần trả lời.

#### Vậy nhóm đã học được gì:
- Khánh Đăng:
    + Hiểu rõ hơn về quy trình khoa học dữ liệu, tầm quan trọng của việc đặt ra câu hỏi có ý nghĩa cần trả lời. 
    + Biết thêm về các hàm vẽ đồ thị và cách phân tích nó.
- Minh Tú:
    + Hiểu hơn về quy trình khoa học dữ liệu.
    + Hiểu hơn về các xử dụng git, xử lý conflict lúc merge. 


#### Nếu có thêm thời gian thì sẽ làm gì?
- Nhóm sẽ tìm thêm câu hỏi có ý nghĩa để trả lời.

### 5. Tài liệu tham kháo: <a class="anchor" id="5"></a>
-  Các slide, video record của môn học.
-  https://stackoverflow.com/ ( các topic về python )